crypto/objects/obj_xref.c

```diff
@@ -40,69 +42,108 @@ static int sigx_cmp(const nid_triple *const *a, const nid_triple *const *b)
 
 IMPLEMENT_OBJ_BSEARCH_CMP_FN(const nid_triple *, const nid_triple *, sigx);
 
+static CRYPTO_ONCE sig_init = CRYPTO_ONCE_STATIC_INIT;
+
+DEFINE_RUN_ONCE_STATIC(o_sig_init)
+{
+    sig_app = sk_nid_triple_new(sig_sk_cmp);
+    sigx_app = sk_nid_triple_new(sigx_cmp);
+    sig_lock = CRYPTO_THREAD_lock_new();
+
+    if (sig_app == NULL || sigx_app == NULL || sig_lock == NULL) {
+        sk_nid_triple_free(sig_app);
+        sk_nid_triple_free(sigx_app);
+        CRYPTO_THREAD_lock_free(sig_lock);
+        sig_app = NULL;
+        sigx_app = NULL;
+        sig_lock = NULL;
+        return 0;
+    }
+    return 1;
+}
+
+static int obj_sig_init(void)
+{
+    return RUN_ONCE(&sig_init, o_sig_init);
+}
+
 int OBJ_find_sigid_algs(int signid, int *pdig_nid, int *ppkey_nid)
 {
     nid_triple tmp;
     const nid_triple *rv = NULL;
+    int idx, lock = sig_app != NULL;
+
     tmp.sign_id = signid;
 
-    if (sig_app != NULL) {
-        int idx = sk_nid_triple_find(sig_app, &tmp);
-        rv = sk_nid_triple_value(sig_app, idx);
+    if (lock) {
+        if (!CRYPTO_THREAD_read_lock(sig_lock))
+            return 0;
+        idx = sk_nid_triple_find(sig_app, &tmp);
+        if (idx >= 0)
+            rv = sk_nid_triple_value(sig_app, idx);
     }
 #ifndef OBJ_XREF_TEST2
     if (rv == NULL) {
+        if (lock) {
+            CRYPTO_THREAD_unlock(sig_lock);
+            lock = 0;
+        }
         rv = OBJ_bsearch_sig(&tmp, sigoid_srt, OSSL_NELEM(sigoid_srt));
     }
 #endif
-    if (rv == NULL)
-        return 0;
-    if (pdig_nid)
-        *pdig_nid = rv->hash_id;
-    if (ppkey_nid)
-        *ppkey_nid = rv->pkey_id;
-    return 1;
+    if (rv != NULL) {
+        if (pdig_nid != NULL)
+            *pdig_nid = rv->hash_id;
+        if (ppkey_nid != NULL)
+            *ppkey_nid = rv->pkey_id;
+    }
+    if (lock)
+        CRYPTO_THREAD_unlock(sig_lock);
+    return rv != NULL;
 }
 
 int OBJ_find_sigid_by_algs(int *psignid, int dig_nid, int pkey_nid)
 {
     nid_triple tmp;
     const nid_triple *t = &tmp;
     const nid_triple **rv = NULL;
+    int idx, lock = sigx_app != NULL;
 
     tmp.hash_id = dig_nid;
     tmp.pkey_id = pkey_nid;
 
-    if (sigx_app) {
-        int idx = sk_nid_triple_find(sigx_app, &tmp);
+    if (lock) {
+        if (!CRYPTO_THREAD_read_lock(sig_lock))
+            return 0;
+        idx = sk_nid_triple_find(sigx_app, &tmp);
         if (idx >= 0) {
             t = sk_nid_triple_value(sigx_app, idx);
             rv = &t;
         }
     }
 #ifndef OBJ_XREF_TEST2
     if (rv == NULL) {
+        if (lock) {
+            CRYPTO_THREAD_unlock(sig_lock);
```

Comment:
Also here - can't we unlock regardless of the value of rv?

Reply:
It depends if `(*rv)->sign_id` can change out of lock.  Currently it can't.

The code cleans up a bit if we forgo this safety -- the unlock can move up to after line 122.

Reply:
As before...we do add only, so I think we can assume that `->sign_id` won't change.